agent/cache/cache.go

```diff
@@ -609,13 +604,91 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	// identical calls to fetch will return the same waiter rather than
 	// perform multiple fetches.
 	entry.Fetching = true
+	entry.HasGoroutine = true
 	c.entries[key] = entry
 	metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
 	metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
 
-	tEntry := r.TypeEntry
 	// The actual Fetch must be performed in a goroutine.
-	go func() {
+	go c.launchBackgroundFetcher(key, r, entry)
+
+	return entry.Waiter
+}
+
+func (c *Cache) launchBackgroundFetcher(key string, r getOptions, initialEntry cacheEntry) {
+	defer c.mutateExistingEntry(key, r, func(entry *cacheEntry) {
+		if entry.HasGoroutine {
+			entry.HasGoroutine = false
+			entry.Fetching = false
+		}
+	})
+
+	var (
+		attempt uint
+		entry   = initialEntry
+	)
+	for {
+		if !c.runBackgroundFetcherOnce(key, r, entry, &attempt) {
```

Comment:
I like that. I mostly was trying to avoid this feeling like the syntax of `append()` where you pass the prior and return the updated and the caller has to do the actual update.